core/src/db/DBImpl.cpp

```diff
@@ -776,6 +783,12 @@ DBImpl::InsertVectors(const std::string& collection_id, const std::string& parti
         status = ExecWalRecord(record);
     }
 
+#ifdef MILVUS_APU_VERSION
+    if (!vectors.binary_data_.empty()) {
+        auto apu = Fpga::ApuInst::getInstance();
+        status = apu->insertVectors(vectors, collection_id);
+    }
+#endif
```

Comment:
Why APU need to do data management by itself?
The usual way is milvus receive data, persist to hard disk, merge segments, build index.  When search request come in,  the special compute unit(such as FPGA) load segments into its memory and compute.

The FPGA implement is in this way, please refer this commit: https://github.com/milvus-io/milvus/pull/3996/files
1. it doesn't change the data management of milvus
2. only rewrite the process when load the data during search

Reply:
Hi, 
loading data to apu is very expensive operation.  to overcome it , we implment an add/remove opertions to save time when collection has been already loaded to the apu. 

in order to make APU-Milvus collection synchronized , when an insert/delete api is called , we are performing the same change in APU device .

Reply:
The apu->insertVectors() is to pass the vectors into APU memory? Then when the server restart and search, how the APU load data from disk storage?

Reply:
Hi , 
insertVectors is having effect only when the dataset is already loaded. when restarting Milvus , Apu is been totaly  cleaned until next search . 

for example :
 user created a dataset of 1M records.
 In the first search the dataset is been loaded to the Apu and search action is performed. 
user added more records to the the same collection , insertVectors is called  and the records is been loaded to the APU together with the collection that already loaded. The total collection is not been loded again . 
all the collection and the added vectors are ready for search. 

restarting Milvus will cause all the loaded data do be cleared and it will be loaded in the next flat search .